plugins/autodim.koplugin/main.lua

```diff
@@ -0,0 +1,257 @@
+--[[--
+Plugin for automatic dimming of the frontlight after an idle period.
+
+@module koplugin.autodim
+--]]--
+
+local Device = require("device")
+local Event = require("ui/event")
+local FFIUtil = require("ffi/util")
+local SpinWidget = require("ui/widget/spinwidget")
+local UIManager = require("ui/uimanager")
+local WidgetContainer = require("ui/widget/container/widgetcontainer")
+local time = require("ui/time")
+local _ = require("gettext")
+local T = FFIUtil.template
+
+local DEFAULT_AUTODIM_STARTTIME_M = 5
+local DEFAULT_AUTODIM_DURATION_S = 20
+local DEFAULT_AUTODIM_FRACTION = 20
+local AUTODIM_EVENT_FREQUENCY = 2 -- in Hz; Frequenzy for FrontlightChangedEvent on E-Ink devices
+
+local AutoDim = WidgetContainer:new{
+    name = "autodim",
+}
+
+function AutoDim:init()
+    self.autodim_starttime_m = G_reader_settings:readSetting("autodim_starttime_minutes", -1)
+    self.autodim_duration_s = G_reader_settings:readSetting("autodim_duration_seconds", DEFAULT_AUTODIM_DURATION_S)
+    self.autodim_fraction = G_reader_settings:readSetting("autodim_fraction", DEFAULT_AUTODIM_FRACTION)
+
+    self.last_action_time = UIManager:getElapsedTimeSinceBoot()
+
+    self.ui.menu:registerToMainMenu(self)
+    UIManager.event_hook:registerWidget("InputEvent", self)
+
+    self:_schedule_autodim_task()
+    self.isCurrentlyDimming = false -- true during or after the dimming ramp
+end
+
+function AutoDim:addToMainMenu(menu_items)
+    menu_items.autodim = self:getAutodimMenu()
+end
+
+function AutoDim:getAutodimMenu()
+    return {
+        text = _("Automatic dimmer"),
+        checked_func = function() return self.autodim_starttime_m > 0 end,
+        sub_item_table = {
+            {
+                text_func = function()
+                    return self.autodim_starttime_m <= 0 and _("Idle time for dimmer") or
+                    T(_("Idle time for dimmer: %1 min"), self.autodim_starttime_m)
+                end,
+                checked_func = function() return self.autodim_starttime_m > 0 end,
+                callback = function(touchmenu_instance)
+                    local idle_dialog = SpinWidget:new{
+                        title_text = _("Automatic dimmer idle time"),
+                        info_text = _("Start the dimmer after the designated period of inactivity."),
```

Comment:
"dimmer" as a noun for the "thing" that handles the feature sounds strange to my non-english ears.
It would feel less strange (to my ears) as a verb: "dimming idle time", "start dimming". But what do I know ? :)

Reply:
It's also the name for the knob thingy that dims a real lightbulb (un *variateur* in French), so it kinda fits.

But, yeah, I get where you're coming from ;).